cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -377,82 +403,19 @@ func (m *McmManager) GetMachineDeploymentForMachine(machine *Ref) (*MachineDeplo
-		}
-
-		str := strings.Split(s.Name, ".")
-		_, Name := str[0], str[1]
-
-		if Name == machineDeploymentName {
-			min = s.MinSize
-			max = s.MaxSize
-			break
-		}
-	}
-
-	return &MachineDeployment{
-		mcmRef,
-		m,
-		min,
-		max,
-	}, nil
+	return machineDeployment, nil
 }
 
 // Refresh method, for each machine deployment, will reset the priority of the machines if the number of annotated machines is more than desired.
 // It will select the machines to reset the priority based on the descending order of creation timestamp.
 func (m *McmManager) Refresh() error {
-	machineDeployments, err := m.machineDeploymentLister.MachineDeployments(m.namespace).List(labels.Everything())
-	if err != nil {
-		klog.Errorf("[Refresh] unable to list machine deployments")
-		return err
-	}
 	var collectiveError error
-	for _, machineDeployment := range machineDeployments {
-		// ignore the machine deployment if it is in rolling update
-		if !isRollingUpdateFinished(machineDeployment) {
-			klog.Infof("[Refresh] machine deployment %s is under rolling update, skipping", machineDeployment.Name)
-			continue
-		}
-		replicas := machineDeployment.Spec.Replicas
-		// check if number of annotated machine objects is more than desired and correspondingly reset the priority annotation value if needed.
-		machines, err := m.getMachinesForMachineDeployment(machineDeployment.Name)
-		if err != nil {
-			klog.Errorf("[Refresh] failed to get machines for machine deployment %s, hence skipping it. Err: %v", machineDeployment.Name, err.Error())
-			collectiveError = errors.Join(collectiveError, err)
-			continue
-		}
-		var machinesMarkedForDeletion []*v1alpha1.Machine
-		for _, machine := range machines {
-			// no need to reset priority for machines already in termination or failed phase
-			if machine.Status.CurrentStatus.Phase == v1alpha1.MachineTerminating || machine.Status.CurrentStatus.Phase == v1alpha1.MachineFailed {
-				continue
-			}
-			if annotValue, ok := machine.Annotations[machinePriorityAnnotation]; ok && annotValue == priorityValueForCandidateMachines {
-				machinesMarkedForDeletion = append(machinesMarkedForDeletion, machine)
-			}
-		}
-		if int(replicas) > len(machines)-len(machinesMarkedForDeletion) {
-			slices.SortStableFunc(machinesMarkedForDeletion, func(m1, m2 *v1alpha1.Machine) int {
-				return -m1.CreationTimestamp.Compare(m2.CreationTimestamp.Time)
-			})
-			diff := int(replicas) - len(machines) + len(machinesMarkedForDeletion)
-			targetRefs := make([]*Ref, 0, diff)
-			for i := 0; i < min(diff, len(machinesMarkedForDeletion)); i++ {
-				targetRefs = append(targetRefs, &Ref{Name: machinesMarkedForDeletion[i].Name, Namespace: machinesMarkedForDeletion[i].Namespace})
-			}
-			collectiveError = errors.Join(collectiveError, m.resetPriorityForMachines(targetRefs))
-		}
+	for _, machineDeployment := range m.machineDeployments {
+		collectiveError = errors.Join(collectiveError, machineDeployment.Refresh())
```

Comment:
non-optimal use of `errors.Join` which will create repeated struct instances. `errors.Join` creates a new error object by combining multiple errors. If used in a loop, it allocates memory each time it's called. The right usage is  first make a `var errs []error` outside the loop, append individual errors  into this slice within the loop and then use `errors.Join` outside the loop to create the final error.